[dtensor] add src_data_rank to distribute_tensor API

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #144005
* __->__ #143883

As titled, this PR add a kwarg src_data_rank to the distribute_tensor
API, to allow user specify a specific rank as the full tensor source
data. Previously we by default specify group_rank=0 as the source of
truth for single device semantic, this new option:

* gives advanced user flexiblity to choose the source data rank
* allow user to specify None explicity, which means we will skip the
  communications needed (scatter/broadcast) for the cases that does not
care about single device semantic (i.e. loading from a checkpoint)

cc @H-Huang @awgu @kwen2501 @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

wanchaol committed:
[dtensor] add src_data_rank to distribute_tensor API

As titled, this PR add a kwarg src_data_rank to the distribute_tensor
API, to allow user specify a specific rank as the full tensor source
data. Previously we by default specify group_rank=0 as the source of
truth for single device semantic, this new option:

* gives advanced user flexiblity to choose the source data rank
* allow user to specify None explicity, which means we will skip the
  communications needed (scatter/broadcast) for the cases that does not
care about single device semantic (i.e. loading from a checkpoint)

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add src_data_rank to distribute_tensor API"


As titled, this PR add a kwarg src_data_rank to the distribute_tensor
API, to allow user specify a specific rank as the full tensor source
data. Previously we by default specify group_rank=0 as the source of
truth for single device semantic, this new option:

* gives advanced user flexiblity to choose the source data rank
* allow user to specify None explicity, which means we will skip the
  communications needed (scatter/broadcast) for the cases that does not
care about single device semantic (i.e. loading from a checkpoint)

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

tianyu-l commented on torch/distributed/tensor/placement_types.py:
```diff
@@ -173,7 +177,11 @@ def _shard_tensor(
 
         mesh_dim_local_rank = my_coordinate[mesh_dim]
         output = torch.empty_like(scatter_list[mesh_dim_local_rank])
```
It looks when `src_data_rank is None`, `_shard_tensor` would return a tensor initiated using `empty_like`. This behavior is a bit different from what I'd expect of "using local data" -- e.g. if `distribute_tensor` is called on a tensor with replicate data on each rank, I would have expected the semantics of `Shard` to be kept even if `src_data_rank is None`. But maybe this is not an important use case.

wanchaol replied:
Good catch! Updated so that it respects the "local data" semantic in the doc :)

tianyu-l commented on torch/distributed/tensor/_collective_utils.py:
```diff
@@ -79,6 +77,8 @@ def mesh_scatter(
     mesh_dim: int = 0,
     async_op: bool = False,
+    *,
+    group_src: int = 0,
```
do you think the doc string is worth updating, too?

wanchaol replied:
done

tianyu-l commented on torch/distributed/tensor/_collective_utils.py:
```diff
@@ -136,6 +131,8 @@ def mesh_broadcast(
+    group_src: int = 0,
 ) -> Optional[Work]:
     """
     broadcast the tensor to a device mesh dimension. We by default
```
ditto

wanchaol replied:
done

tianyu-l reviewed (commented):
Welcome back!

Had a comment on the semantics of `distribute_tensor` with shard placements when `src_data_rank is None`.

wanchaol committed:
Update on "[dtensor] add src_data_rank to distribute_tensor API"


As titled, this PR add a kwarg src_data_rank to the distribute_tensor
API, to allow user specify a specific rank as the full tensor source
data. Previously we by default specify group_rank=0 as the source of
truth for single device semantic, this new option:

* gives advanced user flexiblity to choose the source data rank
* allow user to specify None explicity, which means we will skip the
  communications needed (scatter/broadcast) for the cases that does not
care about single device semantic (i.e. loading from a checkpoint)

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] add src_data_rank to distribute_tensor API"


As titled, this PR add a kwarg src_data_rank to the distribute_tensor
API, to allow user specify a specific rank as the full tensor source
data. Previously we by default specify group_rank=0 as the source of
truth for single device semantic, this new option:

* gives advanced user flexiblity to choose the source data rank
* allow user to specify None explicity, which means we will skip the
  communications needed (scatter/broadcast) for the cases that does not
care about single device semantic (i.e. loading from a checkpoint)

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

XilunWu commented on test/distributed/_tensor/test_api.py:
```diff
@@ -41,15 +45,21 @@ def world_size(self) -> int:
 
     @with_comms
-    def test_distribute_tensor(self):
+    def test_distribute_tensor_rank(self):
```
it would be good if we also test uneven sharding.

wanchaol replied:
I think this feature is somewhat orthogonal to even or uneven sharding, I'll try to update as a follow up later

XilunWu reviewed (approved):
LGTM! Thanks for supporting 0-collective tensor sharding.

tianyu-l reviewed (approved):
LGTM!
